File: superstitioMain/java/superstitio/cards/general/SkillCard/hangUpCard/PassiveGangBang.java
package superstitio.cards.general.SkillCard.hangUpCard;

import com.megacrit.cardcrawl.cards.AbstractCard;
import com.megacrit.cardcrawl.characters.AbstractPlayer;
import com.megacrit.cardcrawl.monsters.AbstractMonster;
import superstitio.DataManager;
import superstitio.cards.general.GeneralCard;
import superstitio.delayHpLose.RemoveDelayHpLoseBlock;
import superstitioapi.hangUpCard.CardOrb;
import superstitioapi.hangUpCard.CardOrb_CardTrigger;
import superstitioapi.hangUpCard.Card_TriggerHangCardManually;

import static superstitioapi.InBattleDataManager.getHangUpCardOrbGroup;
import static superstitioapi.actions.AutoDoneInstantAction.addToBotAbstract;

public class PassiveGangBang extends GeneralCard implements Card_TriggerHangCardManually {
    public static final String ID = DataManager.MakeTextID(PassiveGangBang.class);

    public static final CardType CARD_TYPE = CardType.SKILL;

    public static final CardRarity CARD_RARITY = CardRarity.RARE;

    public static final CardTarget CARD_TARGET = CardTarget.SELF;

    private static final int COST = 2;
    private static final int BLOCK = 14;
    private static final int UPGRADE_BLOCK = 4;
    private static final int MAGIC = 3;
    private static final int UPGRADE_MAGIC = 1;

    public PassiveGangBang() {
        super(ID, CARD_TYPE, COST, CARD_RARITY, CARD_TARGET);
        this.setupBlock(BLOCK, UPGRADE_BLOCK, new RemoveDelayHpLoseBlock());
        this.setupMagicNumber(MAGIC, UPGRADE_MAGIC);
    }

    @Override
    public void use(AbstractPlayer player, AbstractMonster monster) {
        addToBot_gainBlock();
        AbstractCard self = this;
        for (int i = 0; i < this.magicNumber; i++) {
            addToBotAbstract(() -> getHangUpCardOrbGroup()
                    .ifPresent(cardGroup -> cardGroup.cards.stream()
                            .filter(orb -> orb instanceof CardOrb_CardTrigger)
                            .forEach(orb -> orb.forceAcceptAction(self))));
        }
    }

    @Override
    public void upgradeAuto() {
    }

    @Override
    public boolean forceFilterCardOrbToHoveredMode(CardOrb orb) {
//        return orb instanceof CardOrb_CardTrigger;
        return true;
    }

    @Override
    public int forceChangeOrbCounterShown(CardOrb orb) {
//        if (orb instanceof CardOrb_CardTrigger)
        return Math.max(orb.OrbCounter - this.magicNumber, 0);
//        return orb.OrbCounter;
    }
}

File: superstitioMain/java/superstitio/cards/general/SkillCard/hangUpCard/PassiveGangBang.kt
package superstitio.cards.general.SkillCard.hangUpCard

import com.megacrit.cardcrawl.cards.AbstractCard
import com.megacrit.cardcrawl.characters.AbstractPlayer
import com.megacrit.cardcrawl.monsters.AbstractMonster
import superstitio.DataManager
import superstitio.cards.general.GeneralCard
import superstitio.delayHpLose.RemoveDelayHpLoseBlock
import superstitioapi.hangUpCard.*

class PassiveGangBang : GeneralCard(ID, CARD_TYPE, COST, CARD_RARITY, CARD_TARGET), Card_TriggerHangCardManually
{
    init
    {
        this.setupBlock(BLOCK, UPGRADE_BLOCK, RemoveDelayHpLoseBlock())
        this.setupMagicNumber(MAGIC, UPGRADE_MAGIC)
    }

    override fun use(player: AbstractPlayer?, monster: AbstractMonster?)
    {
        addToBot_gainBlock()
        val self: AbstractCard = this
        for (i in 0 until this.magicNumber)
        {
            HangUpCardGroup.forEachHangUpCard { orb: CardOrb ->
                if (orb is ICardOrb_WaitTime || orb is ICardOrb_EachTime)
                    orb.forceAcceptAction(self)
            }.addToBotAsAbstractAction()
        }
    }

    override fun upgradeAuto()
    {
    }

    override fun forceFilterCardOrbToHoveredMode(orb: CardOrb): Boolean
    {
        return orb is ICardOrb_WaitTime || orb is ICardOrb_EachTime
    }

    override fun forceChangeOrbCounterShown(orb: CardOrb): Int
    {
        return if (orb is ICardOrb_WaitTime || orb is ICardOrb_EachTime)
            orb.orbCounter.toInt { it - this.magicNumber }
        else
            orb.orbCounter.toInt()
    }

    companion object
    {
        val ID: String = DataManager.MakeTextID(PassiveGangBang::class.java)

        val CARD_TYPE: CardType = CardType.SKILL

        val CARD_RARITY: CardRarity = CardRarity.RARE

        val CARD_TARGET: CardTarget = CardTarget.SELF

        private const val COST = 2
        private const val BLOCK = 14
        private const val UPGRADE_BLOCK = 4
        private const val MAGIC = 2
        private const val UPGRADE_MAGIC = 1
    }
}
